src/transports/tcp.rs

```diff
@@ -565,3 +549,67 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
         Ok(data)
     }
 }
+
+fn set_tcp_stream_config(stream: &TcpStream, config: &TcpTransportConfig) {
+    if let Err(e) = stream.set_nonblocking(false) {
+        println!("Error setting nonblocking: {:?}", e);
+    }
+    // if let Err(e) = stream.set_linger(Some(config.write_timeout)) {
+    //     println!("Error setting linger: {:?}", e);
+    // }
+    if let Err(e) = stream.set_read_timeout(Some(config.read_timeout)) {
+        println!("Error setting read timeout: {:?}", e);
+    }
+    if let Err(e) = stream.set_write_timeout(Some(config.write_timeout)) {
+        println!("Error setting write timeout: {:?}", e);
+    }
+}
+
+fn write_exact_timeout(
+    endpoint: &mut TcpEndpoint,
+    data: &Vec<u8>,
+    timeout: Option<Duration>,
+) -> PeerNetResult<Duration> {
+    let start_time = Instant::now();
+    let msg_size: u32 = data.len().try_into().map_err(|_| {
+        PeerNetError::SendError.error("error with send len", Some(format!("{:?}", data.len())))
+    })?;
+
+    if msg_size > endpoint.config.max_message_size as u32 {
+        return Err(
+            PeerNetError::SendError.error("send len too long", Some(format!("{:?}", data.len())))
+        );
+    }
+
+    // if we don't have a timeout, use the default write timeout
+    let timeout = timeout.unwrap_or(endpoint.config.write_timeout);
+
+    endpoint
+        .stream
+        .set_write_timeout(Some(timeout))
+        .map_err(|e| {
+            PeerNetError::SendError.error("error setting write timeout", Some(e.to_string()))
+        })?;
+
+    let mut write_count = 0;
+    while write_count < data.len() {
+        if start_time.elapsed() >= timeout {
+            return Err(PeerNetError::SendError.error("send write timeout", None));
+        }
+
+        match endpoint.stream.write(data[write_count..].as_ref()) {
```

Comment:
incorrect implementation: at every iteration the stream timeout should be set to the **remaining** time left until timeout

Reply:
done

Reply:
recommendation:

```rust
    fn read_exact_timeout(stream: &mut TcpStream, buf: &mut[u8], timeout: Duration) -> Result<(), Error> {
        let mut total_read = 0;
        let start_time = Instant::now();
        while total_read < buf.len() {
            let remaining = timeout.saturating_sub(start_time.elapsed());
            if remaining.is_zero() {
                return Err(Error::Timeout);
            }
            stream.set_read_timeout(Some(remaining)).map_err(|e| Error::CouldNotSetTimeout(e))?;
            match stream.read(&mut buf[total_read..]) {
                Err(e) => return Err(Error::ReadError(e)),
                Ok(0) => return Err(Error::ConnectionClosed),
                Ok(read) => total_read += read,
            }
        }
        Ok(())
    }
```